File: fit/trajectory.py
"""Routines for fitting a function $f: \\mathbb{R}^n \\to \\mathbb{R}^m$

This file contains functions for fitting multi-dimensional functions 
of many variables. This is done using the maximum likelyhood principle 
and a "trajectory method":
the function $\vec{y} = \vec{f}(\vec{x})$ is thought as 
a trajectory $\vec{z}(t)$ in a n+m dimensional space: 

$$z_i = x_i \, , \, i=1,...,n$$
$$z_i = y_i \, , \, i=n+1,...,m$$

The fit minimizes the distance of this trajectory 
from the theoretical expectation.
"""

import numpy as np
import scipy.optimize as opt

def fit_trajectory(
    ansatz, 
    x: np.ndarray, ex: np.ndarray, 
    y: np.ndarray, ey: np.ndarray, 
    guess: np.ndarray, 
    method = "BFGS",
    Cov_inv = None
    ):
    """
    Fit a function f: \\mathbb{R}^n \\to \\mathbb{R}^m with the trajectory method

    Args:
        ansatz : function f(x, p): x=1d array, p=1d array, returns 1d array
        x (np.ndarray): x values: array of size (N_pts, n)
        ex (np.ndarray): errors on x values: array of size (N_pts, n)
        y (np.ndarray): y values: array of size (N_pts, m)
        ey (np.ndarray): errors on y values: array of size (N_pts, m)
        guess (np.ndarray): guesses for the ansatz of the fit
        maxiter (int, optional): _description_. Defaults to 10000.
        method (str, optional): _description_. Defaults to "BFGS".
        Cov_inv: inverse covariance matrix estimate (e.g. from a fit on the bootstrap mean)

    Returns:
        dict: Dictionary with the information about the fit
    """
    assert (x.shape[0] == y.shape[0]) ## same number of points
    N_pts = x.shape[0] # number of points
    Nx = x.flatten().shape[0]
    ix_with_err = (ex > 0) # indices of points with error
    x_fit = np.copy(x[ix_with_err])
    ex_fit = ex[ix_with_err]
    iy_with_err = np.where(ey > 0) # indices of points with error
    Ny = y.flatten().shape[0]
    y_fit = y[iy_with_err]
    ey_fit = ey[iy_with_err]

    N_par = len(guess) # number of parameters of the fit ansatz
    N_dof = N_pts - N_par # number of degrees of freedom

    # chi square residual function
    def ch2_uncorr(p_all):
        p_ansatz = p_all[0:N_par] # parameters of the fit only
        p_x = p_all[N_par:] 
        ch2_x = np.sum(((x_fit - p_x)/ex_fit)**2)
        y_th = np.array([ansatz(x[i,:], p_ansatz) for i in range(N_pts)])[iy_with_err] # theoretical values
        # print(".", (y_fit - y_th).flatten()[0:4])
        ch2_y = np.sum(((y_fit - y_th)/ey_fit)**2)
        ch2_res = ch2_x + ch2_y
        return ch2_res
    if Cov_inv is None:
        ch2 = lambda p: ch2_uncorr(p)
    else:
        assert(len(Cov_inv.shape) == 2)
        assert(Cov_inv.shape[0] == Cov_inv.shape[1])
        assert(Cov_inv.shape[0] == Nx+Ny)
        X_th = np.copy(x)
        Y_th = np.copy(y)
        def ch2(p_all):
            p_ansatz = p_all[0:N_par] # parameters of the fit only
            p_x = p_all[N_par:]
            X_th[ix_with_err] = p_x 
            dx = (x - X_th).T
            Y_th[iy_with_err] = np.array([ansatz(X_th[i,:], p_ansatz) for i in range(N_pts)])[iy_with_err]
            dy = (y - Y_th).T
            z = np.concatenate((dx.flatten(), dy.flatten()))
            ch2_res = np.sum(z.T @ Cov_inv @ z)
            # print(",", dy.flatten()[0:4])
            # ch2_uncorr_value = ch2_uncorr(p_all=p_all)
            return ch2_res
    #-------
    guess = np.concatenate((guess, np.copy(x[ix_with_err].flatten())))
    mini = opt.minimize(fun = ch2, x0 = guess, method = method)
    par = mini.x
    ch2_value = ch2(par)
    
    res = dict({})

    res["ansatz"] = ansatz
    res["N_par"] = N_par
    res["N_pts"] = N_pts
    res["par"] = par
    res["ch2"] = ch2_value
    res["N_dof"] = N_dof ## degrees of freedom
    ch2_dof = float("nan")
    if N_dof > 0:
        ch2_dof = ch2_value / N_dof
    #---
    res["ch2_dof"] = ch2_dof
    return(res)
#---

File: fit/trajectory.cpp
#include <nanobind/nanobind.h>
#include <nanobind/eigen/dense.h>
#include <nanobind/stl/function.h>
#include <nanobind/stl/string.h>

#include <iostream>
#include <Eigen/Dense>
#include <cmath>
#include <functional>
#include <limits>
#include <stdexcept>
#include <vector>

namespace nb = nanobind;

// Define matrix/vector types using Eigen
using Matrix = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
using Vector = Eigen::VectorXd;

nb::dict fit_trajectory(
    std::function<Vector(const Vector&, const Vector&)> ansatz,
    const Matrix& x,
    const Matrix& ex,
    const Matrix& y,
    const Matrix& ey,
    const Vector& guess,
    const std::string& method = "BFGS",
    const nb::object& cov_inv_obj = nb::none()
) {

  std::cout << "c++ fitting routine with nanobind!\n";
  
    if (x.rows() != y.rows()) {
        throw std::invalid_argument("x and y must have the same number of rows.");
    }

    const int64_t N_pts = x.rows();
    const int64_t Nx_cols = x.cols();
    const int64_t Ny_cols = y.cols();
    const int64_t N_par = guess.size();
    const int64_t N_dof = N_pts - N_par;

    // Identify indices where ex > 0 and ey > 0
    std::vector<std::pair<int64_t, int64_t>> ix_with_err;
    for (int64_t r = 0; r < x.rows(); ++r) {
        for (int64_t c = 0; c < x.cols(); ++c) {
            if (ex(r, c) > 0.0) {
                ix_with_err.emplace_back(r, c);
            }
        }
    }

    std::vector<std::pair<int64_t, int64_t>> iy_with_err;
    for (int64_t r = 0; r < y.rows(); ++r) {
        for (int64_t c = 0; c < y.cols(); ++c) {
            if (ey(r, c) > 0.0) {
                iy_with_err.emplace_back(r, c);
            }
        }
    }

    // Build flattened vector of guess parameters: [p_ansatz, p_x_initial]
    Vector full_guess(N_par + ix_with_err.size());
    full_guess.head(N_par) = guess;
    for (size_t k = 0; k < ix_with_err.size(); ++k) {
        auto [r, c] = ix_with_err[k];
        full_guess(N_par + k) = x(r, c);
    }

    // Determine whether Cov_inv was passed
    bool has_cov_inv = !cov_inv_obj.is_none();
    Matrix cov_inv;

    if (has_cov_inv) {
        cov_inv = nb::cast<Matrix>(cov_inv_obj);
        int64_t total_dim = x.size() + y.size();
        if (cov_inv.rows() != cov_inv.cols() || cov_inv.rows() != total_dim) {
            throw std::invalid_argument("Cov_inv must be a square matrix of size (Nx + Ny, Nx + Ny)");
        }
    }

    // Chi-squared objective function evaluation
    auto eval_chi2 = [&](const Vector& p_all) -> double {
        Vector p_ansatz = p_all.head(N_par);

        if (!has_cov_inv) {
            double ch2_x = 0.0;
            for (size_t k = 0; k < ix_with_err.size(); ++k) {
                auto [r, c] = ix_with_err[k];
                double val = (x(r, c) - p_all(N_par + k)) / ex(r, c);
                ch2_x += val * val;
            }

            // Construct trajectory evaluation X_th
            Matrix X_th = x;
            for (size_t k = 0; k < ix_with_err.size(); ++k) {
                auto [r, c] = ix_with_err[k];
                X_th(r, c) = p_all(N_par + k);
            }

            double ch2_y = 0.0;
            for (int64_t i = 0; i < N_pts; ++i) {
                Vector x_row = X_th.row(i);
                Vector y_th = ansatz(x_row, p_ansatz);

                for (int64_t c = 0; c < Ny_cols; ++c) {
                    if (ey(i, c) > 0.0) {
                        double val = (y(i, c) - y_th(c)) / ey(i, c);
                        ch2_y += val * val;
                    }
                }
            }
            return ch2_x + ch2_y;
        } else {
            Matrix X_th = x;
            for (size_t k = 0; k < ix_with_err.size(); ++k) {
                auto [r, c] = ix_with_err[k];
                X_th(r, c) = p_all(N_par + k);
            }

            Matrix Y_th(N_pts, Ny_cols);
            for (int64_t i = 0; i < N_pts; ++i) {
                Vector x_row = X_th.row(i);
                Y_th.row(i) = ansatz(x_row, p_ansatz);
            }

            // Build dx and dy flattened vectors column-major/transposed matching NumPy logic
            Matrix dx = (x - X_th).transpose();
            Matrix dy = (y - Y_th).transpose();

            Vector z(dx.size() + dy.size());
            z.head(dx.size()) = Eigen::Map<const Vector>(dx.data(), dx.size());
            z.tail(dy.size()) = Eigen::Map<const Vector>(dy.data(), dy.size());

            return z.transpose() * cov_inv * z;
        }
    };

    // Callback to Python scipy.optimize for numerical minimization
    nb::object opt = nb::module_::import_("scipy.optimize");
    nb::object py_eval_chi2 = nb::cpp_function(eval_chi2);

    nb::object mini = opt.attr("minimize")(
        py_eval_chi2,
        full_guess,
        nb::arg("method") = method
    );

    Vector opt_par = nb::cast<Vector>(mini.attr("x"));
    double ch2_value = eval_chi2(opt_par);

    double ch2_dof = (N_dof > 0) ? (ch2_value / static_cast<double>(N_dof)) : std::numeric_limits<double>::quiet_NaN();

    // Prepare response dictionary matching original structure
    nb::dict res;
    res["ansatz"] = ansatz;
    res["N_par"] = N_par;
    res["N_pts"] = N_pts;
    res["par"] = opt_par;
    res["ch2"] = ch2_value;
    res["N_dof"] = N_dof;
    res["ch2_dof"] = ch2_dof;

    return res;
}

NB_MODULE(fit_module, m) {
    m.def("fit_trajectory", &fit_trajectory,
        nb::arg("ansatz"),
        nb::arg("x"),
        nb::arg("ex"),
        nb::arg("y"),
        nb::arg("ey"),
        nb::arg("guess"),
        nb::arg("method") = "BFGS",
        nb::arg("Cov_inv") = nb::none(),
        R"pbdoc(
            Fit a function f: R^n -> R^m with the trajectory method using nanobind and Eigen.
        )pbdoc"
    );
}

